### PR TITLE
Merge 2.9

### PR DESCRIPTION
**Pull request by wallyworld**
Merge 2.9

#13201 Use separate matching functions for block and filesystem volume devices
#13202 Update loggo version in go.mod
#13198 Fix agent binary downloading to only download once.
#13090 Fix refresh when upgrading from pod-spec to sidecar charm
#13204 Remove the now-unused ApplicationWatcherFacade
#13207 When there's an error connecting to mongo, ensure we restart it before trying again

Conflicts mostly imports charm v8 vs v9

```
# Conflicts:
#       api/caasfirewaller/client.go
#       api/caasfirewaller/client_test.go
#       apiserver/common/applicationwatcher.go
#       apiserver/common/applicationwatcher_test.go
#       apiserver/facades/client/application/application.go
#       apiserver/facades/client/charms/client.go
#       apiserver/facades/controller/caasfirewaller/state.go
#       apiserver/facades/controller/caasoperatorprovisioner/state.go
#       apiserver/facades/controller/caasunitprovisioner/state.go
#       apiserver/gui.go
#       scripts/win-installer/setup.iss
#       snap/snapcraft.yaml
#       state/application_test.go
#       version/version.go
#       worker/caasapplicationprovisioner/application_test.go
#       worker/caasapplicationprovisioner/mocks/facade_mock.go
#       worker/caasapplicationprovisioner/worker.go
#       worker/caasfirewallersidecar/applicationworker.go
#       worker/caasfirewallersidecar/applicationworker_test.go
#       worker/caasfirewallersidecar/mocks/logger_mock.go
#       worker/caasoperatorprovisioner/worker.go
```

## QA steps

See PRs

### CODE REVIEW

**Comment by wallyworld** [comment]
[merge 2.9 -> develop](https://trello.com/c/zOASME6E/1689-merge-29-develop)

File: apiserver/facades/client/application/application.go [hunk 1]
```diff
@@ -1053,9 +1058,92 @@ func (api *APIBase) applicationSetCharm(
...
+			return errors.Errorf("cannot upgrade from v1 %s deployment to v2", deployment.DeploymentType)
+		}
+
+		modelConfig, err := api.model.ModelConfig()
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		var supported []string
+		for _, base := range newCharm.Manifest().Bases {
```
Comment (SimonRichardson) [inline apiserver/facades/client/application/application.go:1088]: This can return nil in some situations. We should guard against panics here.

**Review by SimonRichardson — APPROVED** [review]
Approved. Although we should fix this asap - https://github.com/juju/juju/pull/13208#pullrequestreview-720086757

File: apiserver/facades/client/application/application.go (hunk 1 above)
Comment (wallyworld) [reply apiserver/facades/client/application/application.go:1088]: But this is inside an if statement which checks `charm.MetaFormat(newCharm) >= charm.FormatV2` which guarantees Manifest() is not nil right?

**Comment by wallyworld** [comment]
$$merge$$

**Pull request merged**